src/inmanta/deploy/scheduler.py

```diff
@@ -848,46 +1025,54 @@ async def get_resource_intent(self, resource: ResourceIdStr) -> Optional[Resourc
                 return None
             return ResourceIntent(model_version=self._state.version, details=resource_details, dependencies=None)
 
-    async def deploy_start(self, resource: ResourceIdStr) -> Optional[ResourceIntent]:
+    async def deploy_start(self, action_id: uuid.UUID, resource: ResourceIdStr) -> Optional[ResourceIntent]:
         async with self._scheduler_lock:
             # fetch resource details under lock
             resource_details = self._get_resource_intent(resource)
             if resource_details is None or self._state.resource_state[resource].blocked is BlockedStatus.YES:
+                # We are trying to deploy a stale resource.
                 return None
             dependencies = await self._get_last_non_deploying_state_for_dependencies(resource=resource)
             self._deploying_latest.add(resource)
-            return ResourceIntent(model_version=self._state.version, details=resource_details, dependencies=dependencies)
+            resource_intent = ResourceIntent(
+                model_version=self._state.version, details=resource_details, dependencies=dependencies
+            )
+            # Update the state in the database.
+            await self.send_in_progress(action_id, Id.parse_id(ResourceVersionIdStr(f"{resource},v={self._state.version}")))
+            return resource_intent
 
-    async def deploy_done(
-        self,
-        resource: ResourceIdStr,
-        *,
-        attribute_hash: str,
-        resource_state: const.HandlerResourceState,
+    async def deploy_done(self, attribute_hash: str, result: DeployResult) -> None:
+        deployment_result: DeploymentResult = DeploymentResult.from_handler_resource_state(result.resource_state)
+        try:
+            await self.update_scheduler_state_for_finished_deploy(attribute_hash, result, deployment_result)
+        except StaleResource:
+            # The resource is no longer managed, no need to update the database state.
+            pass
+        else:
+            # Write deployment result to the database.
+            await self.send_deploy_done(attribute_hash, result, deployment_result)
```

Comment:
I mentioned in the office that I thought there is a risk in writing this outside of the scheduler lock. But the motivation I had was related to `_task_done()`, which is in practice only called *after* the `deploy_done` call finishes. So it might be fine (and preferable) after all. Only reason I'm still in doubt is that I seem to recall my confidence that it had to be under the lock. Will get back to this.

Reply:
I think the initial motivation for doing it under the lock had to do with the write of the compliance status that might conflict with the write of a new scheduler state version. But iirc (I haven't reviewed that part yet), we discussed this and concluded that we'd write the last released and last deployed attribute hashes, from which the staleness of compliance status could be derived at recovery time. So if that is indeed the case, then I think that we can't conflict with new version and we can't conflict with other resources so then I don't think we need to do this under the lock.